File: Imooc Video Downloader/core/common/fetcher.js
var https = require('https');
var http = require('http');
const DEFAULT_HEADER = { 'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36' };

function Fetcher({ headers } = {}) {
    if (!(this instanceof Fetcher))
        return new Fetcher(headers);

    this.time = 3;
    this._time = 3;
    this.delay = 1000;

    if (headers)
        this.setHeader(headers);
}

Fetcher.prototype.httpGet = function (url, {onDone}) {
    var self = this;
    var _http = url.startsWith("https:") ? https : http;
    _http.get(url, {
        'headers': self.headers
    }, function (res) {

        const { statusCode } = res;
        const content_length = res.headers["content-length"];
        if (statusCode == "403") {
            onDone("403");
            return;
        }
        var length = 0, chunks = [];

        res.on("data", function (chunk) {
            chunks.push(chunk);
            length += chunk.length;
        });
        res.on("end", function () {
            onDone(null, Buffer.concat(chunks, length));
        });


    }).on('error', (err) => {
        if (self.time > 1) {
            self.time--;
            setTimeout(function () {
                self.httpGet(url);
            }, self.delay);
        } else
            onDone(err);
    });


    return self;
}
Fetcher.prototype.setHeader = function (headers) {
    this.headers = {
        ...DEFAULT_HEADER,
        ...headers
    };
    return this;
}
Fetcher.prototype.setRetry = function (time, delay) {
    this.time = time;
    this._time = time;
    this.delay = delay;
    return this;
}
module.exports = Fetcher;

File: XiGua Video Downloader/rust/src/fetcher.rs
use hyper::{Client, Request, Body, Method,StatusCode};
use hyper_tls::HttpsConnector;
use hyper::client::HttpConnector;
use failure::{Error, err_msg};
use bytes::Bytes;

static USER_AGENT: &str ="Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36";

pub struct Fetcher {
    client: Client<HttpsConnector<HttpConnector>>,
}

impl Fetcher {
    pub fn new() -> Fetcher {
        let client =
            Client::builder().build::<_, hyper::Body>(HttpsConnector::new());
        Fetcher {
            client,
        }
    }
    fn create_request(method: Method, url: String) -> hyper::http::Result<Request<Body>> {
        return Request::builder()
            .method(method)
            .uri(url)
            .header("user-agent", USER_AGENT)
            .body(Body::empty());
    }
}

impl Fetcher {
    pub async fn head(&self, url: String) -> Result<usize, Error> {
        let req = Self::create_request(
            Method::HEAD, url
        )?;
        let response = (&self.client).request(req).await?;
        if response.status()==StatusCode::OK{
            let headers = response.headers();
            if headers.contains_key("content-length") {
                let length = headers.get("content-length").unwrap().to_str().unwrap();
                let total = length.parse::<usize>().unwrap();
                return Ok(total);
            }
        }
        Err(err_msg("content-length not find"))
    }

    pub async fn range(&self, url: &String, start: usize, end: usize) -> Result<Bytes, Error> {
        let req = Self::create_request(
            Method::GET,
            format!("{}&range={}-{}", url, start, end)
        )?;
        let response = (&self.client).request(req).await?;
        if response.status()==StatusCode::PARTIAL_CONTENT {
            let body = hyper::body::to_bytes(response.into_body()).await?;
            return Ok(body);
        }
        Err(err_msg("get range error"))
    }
}